src/lib/hierarchy-operations/backend.js

```diff
@@ -0,0 +1,100 @@
+const _ = require('lodash');
+const lineageManipulation = require('./lineage-manipulation');
+
+const HIERARCHY_ROOT = 'root';
+const BATCH_SIZE = 10000;
+
+/*
+Fetches all of the documents associated with the "contactIds" and confirms they exist.
+*/
+async function contactList(db, ids) {
+  const contactDocs = await db.allDocs({
+    keys: ids,
+    include_docs: true,
+  });
+
+  const missingContactErrors = contactDocs.rows.filter(row => !row.doc).map(row => `Contact with id '${row.key}' could not be found.`);
+  if (missingContactErrors.length > 0) {
+    throw Error(missingContactErrors);
+  }
+
+  return contactDocs.rows.reduce((agg, curr) => Object.assign(agg, { [curr.doc._id]: curr.doc }), {});
+}
+
+async function contact(db, id) {
+  try {
+    if (id === HIERARCHY_ROOT) {
+      return undefined;
+    }
+
+    return await db.get(id);
+  } catch (err) {
+    if (err.name !== 'not_found') {
+      throw err;
+    }
+
+    throw Error(`Contact with id '${id}' could not be found`);
+  }
+}
+
+/*
+Given a contact's id, obtain the documents of all descendant contacts
+*/
+async function descendantsOf(db, contactId) {
+  const descendantDocs = await db.query('medic/contacts_by_depth', {
+    key: [contactId],
+    include_docs: true,
+  });
+
+  return descendantDocs.rows
+    .map(row => row.doc)
+    /* We should not move or update tombstone documents */
+    .filter(doc => doc && doc.type !== 'tombstone');
+}
+
+async function reportsCreatedByOrAt(db, createdByIds, createdAtId, skip) {
+  const createdByKeys = createdByIds.map(id => [`contact:${id}`]);
+  const createdAtKeys = createdAtId ? [
+    [`patient_id:${createdAtId}`],
+    [`patient_uuid:${createdAtId}`],
+    [`place_id:${createdAtId}`],
+    [`place_uuid:${createdAtId}`]
+  ] : [];
+
+  const reports = await db.query('medic-client/reports_by_freetext', {
+    keys: [
+      ...createdByKeys,
+      ...createdAtKeys,
+    ],
+    include_docs: true,
+    limit: BATCH_SIZE,
+    skip,
+  });
+
+  return _.uniqBy(reports.rows.map(row => row.doc), '_id');
```

Comment:
* my personal view is javascript needs to be paired with something like lodash to be a usable convenient language
* lodash is almost free since it is in package-lock because of other packages which use it
* i've removed lodash from this PR. but there are other PRs that had one usecase also and did the same. [example](https://github.com/medic/cht-conf/pull/349#discussion_r463294059).